### PR TITLE
Add artificial intelligence (AI) player in Javascript

### PR DESCRIPTION
**Pull request by Khufu-I**
This is a copy of Kotlin AIPlayer but copied to Typescript from #8

AI uses the minimax algorithm which assumes the maximizing player picks the move with the highest score and the opposing player picks the move with the lowest score. Additionally the alpha-beta search optimization is used to eliminate branches that cannot influence the final decision.

This requires a player enum or boolean
```ts
export enum Player {X, O}
```

and a Game class with at least the following properties
```ts
class Game { 
  moves;
  player;
  board;

  get isGameOver(){}
  get status(){}
  executeMove(Move){}
  cloneDeep(){}
}
```

### CODE REVIEW

**Comment by Khufu-I** [comment]
Not sure if this will actually work since the missing pieces are required particularly `Game` state needs to be extracted.

File: src/routes/TicTacToe/AIPlayer.ts [hunk 1]
```diff
@@ -74,10 +67,10 @@ private fun alphabeta(
 /**
  * Difficulty of artificial intelligence (AI) player
  */
-enum class Difficulty(val depth: Int) {
-  Easy(2),
-  Medium(3),
-  Hard(5)
+enum Difficulty {
```
Comment (Vithop) [inline src/routes/TicTacToe/AIPlayer.ts]: It is recomended to not use typescript enums as it has a few pitfalls and doesn't convert to JS very well. Please use const objects instead https://www.typescriptlang.org/docs/handbook/enums.html#objects-vs-enums

File: src/routes/TicTacToe/AIPlayer.ts [hunk 2]
```diff
@@ -39,24 +32,24 @@ private fun evaluationFunction(game: Game, maximizingPlayer: Player) = when (gam
...
   }
 
-  val isSamePlayer = game.player == maximizingPlayer
-  var result = if (isSamePlayer) NEGATIVE_INFINITY else POSITIVE_INFINITY
+  const isSamePlayer = game.player == maximizingPlayer
+  let result = if (isSamePlayer) NEGATIVE_INFINITY else POSITIVE_INFINITY
   for (move in moves) {
     if (newAlpha >= newBeta) break
-    val nextGame = game.deepCopy()
+    const nextGame = game.deepCopy()
```
Comment (Vithop) [inline src/routes/TicTacToe/AIPlayer.ts]: For deepCopy you would need to use [structuredClone](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone) instead but we need to ensure the object is serializable.

File: src/routes/TicTacToe/AIPlayer.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+ * Calculate score of the Game for a player
+ *
+ * @param[game]              The game for which to calculate the score
+ * @param[maximizingPlayer]  The player
+ *
+ * @return     Score of the current Game {score ∈ ℚ}
+ */
+private fun evaluationFunction(game: Game, maximizingPlayer: Player) = when (game.status) {
```
Comment (Vithop) [inline src/routes/TicTacToe/AIPlayer.ts]: This AIPlayer file should go into the lib folder. The Routes folder is a special folder which auto creates navigation routes for the web app

File: src/routes/TicTacToe/AIPlayer.ts (hunk 2 above)
Comment (Khufu-I) [reply src/routes/TicTacToe/AIPlayer.ts]: Very interesting. So this implementation leaves it up to the game object to determine how to implement `deepCopy`, unfortunately `structuredClone` doesn't work with private class member variables.

File: src/routes/TicTacToe/AIPlayer.ts (hunk 1 above)
Comment (Khufu-I) [reply src/routes/TicTacToe/AIPlayer.ts]: Good point

**Review by Vithop — APPROVED** [review]

**Pull request merged**